### ISSUE TITLE
export charts as image

### ISSUE THREAD

**Issue by dncpax** [metabase/metabase#4701]
Hi there. One of the most common use cases is to include charts in documents or emails. I was very surprised to see metabase does not have this capability. Maybe I'm not seeing this button?

If someone invests time into getting a pretty good chart/graph then it's because users will find it useful enough to actually use it. Downloading json or csv will force every user to recreate the chart over and over again in the organization.

 I couldn't find any other ticket about this so created this one.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by tlrobinson** [comment]
Related to #3847

Basically this is a bit difficult for us to implement because we don't have a sufficiently modern browser rendering engine available on the backend to do the rendering to image or PDF.  It's possible we could do it on the frontend though.

**Comment by dncpax** [comment]
@tlrobinson Thanks for taking notice. It's really a necessity. This is the first question I get when pitching metabase to our users. It takes away all the great work done in metabase just because of this simple thing, which I get is not easy at all to implement.

I do not know metabase libs, but there seems to be some options to do this client side. Maybe https://github.com/niklasvh/html2canvas? I wish I could help more... if you need testing I'll gladly help.

**Comment by tlrobinson** [comment]
A possible alternative to doing backend rendering is to use something like html2canvas. Here's a quick proof of concept:

```diff
diff --git a/frontend/src/metabase/query_builder/components/QueryDownloadWidget.jsx b/frontend/src/metabase/query_builder/components/QueryDownloadWidget.jsx
index 423edf6..033115d 100644
--- a/frontend/src/metabase/query_builder/components/QueryDownloadWidget.jsx
+++ b/frontend/src/metabase/query_builder/components/QueryDownloadWidget.jsx
@@ -43,6 +43,7 @@ const QueryDownloadWidget = ({ className, card, result, uuid, token }) =>
                     :
                       null
                 )}
+                <ImageButton type="png" card={card} result={result} className="mr1 text-uppercase text-default" />
             </div>
         </div>
     </PopoverWithTrigger>
@@ -94,4 +95,23 @@ QueryDownloadWidget.propTypes = {
     uuid: PropTypes.string
 };
 
+import html2canvas from "html2canvas";
+import Button from "metabase/components/Button";
+
+const ImageButton = ({ type = "png" }) =>
+    <Button
+        icon="downarrow"
+        onClick={() => {
+            document.querySelector("#react_qb_viz .Visualization").style.backgroundColor = "white";
+            html2canvas(document.querySelector("#react_qb_viz .Visualization")).then(function(canvas) {
+                let a = document.createElement("a");
+                a.href = canvas.toDataURL(`image/${type}`);
+                a.download = "test.png";
+                a.click();
+            });
+        }}
+    >
+        {type}
+    </Button>
+
 export default QueryDownloadWidget;
```

However, it's not perfect. SVG doesn't seem to have CSS applied. This:

![screenshot 2017-04-05 11 10 13](https://cloud.githubusercontent.com/assets/18193/24719970/77a565bc-19f0-11e7-899c-b806439fd38b.png)

Becomes this:

![test 8](https://cloud.githubusercontent.com/assets/18193/24719982/88536094-19f0-11e7-9efd-4848f37f8b05.png)

It might be possible to workaround by inlining all the style though https://github.com/niklasvh/html2canvas/issues/1011

**Comment by sunshineo** [comment]
I believe a lot of people need this feature badly. My current workaround as a programmer would be use chrome headless browser https://developers.google.com/web/updates/2017/04/headless-chrome to open the public link of the chart and save it as a image/pdf.

Several issues with this approach:
1. You need to know programming
1. You need separate computing resource to run the program
1. The "Question/Card" need to be public
1. The email integration is wasted. You have to email yourself after save.

**Comment by wallce** [comment]
suggest export from svg to html just like "https://www.highcharts.com/".

**Comment by benjR** [comment]
Is it possible now that metabase is generating images for Slack, emails... ?

**Comment by robert-claypool** [comment]
html2canvas has several unsupported CSS rules, and if the element is `<svg>`, CSS must be in-lined and icons must be base64 encoded, see the "onclone" callback at https://github.com/robert-claypool/html2canvas-sandbox/blob/master/main.js

https://github.com/cburgmer/rasterizeHTML.js produces better out of-the-box screenshots in my experience.

**Comment by thucnc** [comment]
Another workaround is to share the card public and then use `https://pptraas.com/` to get the image from the public url of the card

**Comment by dacopan** [comment]
another option is to send svg code  of the graph created in the client and in the server render it with some engine. QT for example

**Comment by gmisterk** [comment]
Hi ,
Why not use this library : https://apexcharts.com ?

Or use other javascript  library to generate embeded charts ?

**Comment by akalamoyo** [comment]
or you can just take a snapshot of your screen and crop out your graph

**Comment by SetSails** [comment]
Hi guys, 

we need an ability to easily print entire dashboards as well as entire tables (all displayed rows) including the conditional formatting etc. - just like it is presented on the screen.

If printing is an itch, a PDF export would obviously also work.

Is there any chance this can be implemented?

Thank you & Best Regards
Fabian

**Comment by lins05** [comment]
+1 for the idea of using headless chrome to do the work.

Metabase server could launch a (cached) headless chrome instance, give it the link for the current query (I think it captures all the query details in the url fragment, right?), and asks headless chrome to render it to PDF. We also need a mechanism to bypass the authentication step so the headless chrome instance could visit whatever url on the server (maybe a one-time token?)

The only problem is how to interact with headless chrome from within clojure/java. There are mature drivers like puppeteer for js and pypuppeteer for python, but not sure if there is a good library for this on JVM.

**Comment by msdcanderson** [comment]
Our workaround uses puppeteer to generate a PDF.  The problem we're having is that the PDF returns as a single page - so printing looks poor.  If we tell puppeteer to use multiple pages, then charts are cut.  Not ideal - but worth bearing in mind...

**Comment by lins05** [comment]
@msdcanderson so it's a separate process not managed by metabase itself? I wonder how do you integrate with the metabase web app?

**Comment by msdcanderson** [comment]
@lins05 that's right - we embed the metabase dashboard using the standard metabase features in a website and have an 'export' button that kicks off a script that uses puppeteer to generate the pdf

**Comment by pPanda-beta** [comment]
Metabase already can render html to png

https://github.com/metabase/metabase/blob/134b1e41d507c2e824ac0f0ba1992047e9302bdc/src/metabase/pulse/render/png.clj#L91-L94

What about utilising it?

This is heavily used by metabot to send slack messages with card images.

**Comment by flamber** [comment]
@pPanda-beta Because the Pulse rendering is currently fairly simplistic, so it wouldn't return the same visualization style as what people are viewing and would probably cause even more confusion - see #5493

**Comment by likeshumidity** [comment]
My organization [Inciter](https://www.inciter.io) is considering contributing to this project by addressing this issue. In reviewing the issue, I'm seeing several potential features/expectations outlined below.

## Features
Are any of these more or less valuable to others? We particularly value A and D and would prioritize those. In reviewing others' responses, I'm also seeing potential features that rely on an additional process, and I'm excluding those for the time being as I'm not sure how that would fit into Metabase's operational plans.

### A. Saving Dashboard as PDF
- Preference to maintain layout similar to on screen
- Printing to PDF in browser almost works for this but due to limitations of paper size, etc. layout may change
- Charts and tables may span several pages, would be helpful to have clean page breaks and avoid truncating a chart or table
- Some charts are not fully visible based on Chrome's rending and print preview
- May be able to update CSS media queries for print to PDF with low effort to achieve clean results
- Dashboard filters with nothing selected can be confusing on the PDF without the context of interactivity

### B. Saving a chart or table as a PDF
- Printing to PDF in browser currently works well for this, but it will likely fill a page with the chart (not sure if this is desired behavior)

### C. Save a chart or table as a PNG
- Only area viewable on screen
- Would work like Plotly's feature to `Download plot as png`
![image](https://user-images.githubusercontent.com/8808703/90520897-e8003500-e137-11ea-94da-29c487bcc9de.png)
- Is there a concern about the PNG resolution having a similar issue to making a screenshot? That is, it won't scale well and may be pixelated or blurry.

### D. Save a chart or table as a SVG
- Haven't seen this requested, but this is our preference
- Won't appear pixelated/blurry at any scale
- Uses existing canvas for charts, so doesn't need conversion; not sure how tables should be handled/rendered since not using canvas
- Not sure if background should be white or transparent?
- Microsoft Word since version 2016 allow SVG embedding, Adobe Acrobat and other Adobe products can read/embed SVG as well

**Comment by mijamo** [comment]
I think B is the least interesting here. If you have C or D it makes B very trivial to do manually, also I don't really see a good use case for it (but I might be completely wrong of course).

C. is basically like a "print screen" function, which is usually trivially available in most OS now. It would still be a valuable UX improvement (no need to hit hit the print screen function or to select the right area) but it is still very limited in terms of resolution and the cost of implementing that function might be high compared to the benefits.

I think D. is the most versatile as once you have that you have a good basis to do whatever you want. The only downside I see is that Google Slides/Writer do not support SVG, but it shouldn't block metabase from using it. Apart from the table problem I think the size of elements is the main thing to consider here. I often want the have a certain size, both in terms of ratio and font size, none of those can be configured yet so they depend of your current window size (basically a small window would mean bigger font size as an SVG image). It is especially annoying when working with a big screen and small font, then you need to resize your window a lot to get the right presentation aspect. I think when exporting to SVG it would be great to choose the output size (I think it can be fine to accept that the font is stable, so a bigger output size will mean proportionally smaller fonts, in order not to increase the scope of this issue).

A transparent background would be my preference, you don't know what background is desired and it is pretty trivial to add a background to any image in any software afterwards, whereas removing a background is far from trivial, or even impossible in many software.

**Comment by dncpax** [comment]
Hi there! Very glad to see your proposal!! My 2 cents are: whatever works best for majority of users. That alone excludes screen-print-crop suggestions. My most common use case is saving to image and inserting into email/word. Copy/paste also works but many users seem confused by not having a physical file to move around. I was a bit worried seeing svg as the preferred format but since it is adopted by office tools / email clients it would be better than png.

Thanks again, and look forward to trying that out!

**Comment by alexfrench** [comment]
Sounds good and SVG okay by me. Any progress on this?

**Comment by aaugustin** [comment]
Until we have a built-in solution, which would be fantastic, the best workaround I found with Chrome is:

- Right-click the graph => Inspect element (this opens Developer tools)
- (optional) Enable responsive mode (second icon in the top row of Developer tools) to give the window an appropriate and deterministic size (so I always capture an image of the same size)
- Right-click the `<svg>` node, which is preselected => Capture node screenshot

This downloads a PNG file.

**Comment by aaugustin** [comment]
Also, SVG Crowbar would be great, but it didn't work for me because of the Content Security Policy.

**Comment by neilrobert** [comment]
Hi all
I am embedding Metabasee dashboards into a Webapp
Download to PDF - preferably whole dashboard as it appears on the screeen 
OR - download to PDF - each individual question 
Any idea when this may be available ?

I tried Chrome. - print to PDF - but it resized each question to the same sieze - about a 1/3 of page and only 1/2 wide - each of my small boxes (i. single figure) became big, and the graph is muhc smaller
 
@aaugustin - I tried your Chrome workaround - got developer tools 
"Elements" tab
couldnt find the enablee responsivee mode (you said optional)
OR the 
![Screen Shot 2020-12-03 at 4 49 08 pm](https://user-images.githubusercontent.com/74808162/100969638-80a3af00-3587-11eb-8dda-b397572b97fe.png)
<svg> node to right click on ? - see screen shot

**Comment by blongden** [comment]
@likeshumidity Hi Jacob, how did you get one with this one? it's something I am investigating at the moment too.

**Comment by koe-n** [comment]
Is this feature foreseen for v0.39 as stated in the project roadmap? It don't see it in the list of issues for that milestone.

**Comment by flamber** [comment]
@koe-n No, this will not be in 0.39 - there's currently no timeline on this issue.

**Comment by mahendra350** [comment]
Is there any alternate option to export png and pdf?

**Comment by pPanda-beta** [comment]
Looks like people have found alternatives to get their work done.

**Comment by dncpax** [comment]
just to put here some appreciation for mb...
seeing all the nice stuff mb has been busy doing I almost consider closing this ticket!! it seems small thing in comparison.

having said that, i would still love this to be implemented... but happy to wait.

**Comment by petervuyk456** [comment]
Hi all! @likeshumidity and I are planning to begin work on this issue. Specifically, we are planning to start with feature D (Save chart or table as SVG) as outlined in the prior [comment](https://github.com/metabase/metabase/issues/4701#issuecomment-675496040).

This will be our first contribution to Metabase - as such, we are open to and are actively looking for support in whatever manner it is willing to be provided i.e. strategy, development, etc.

@mijamo @dncpax @alexfrench @blongden @aaugustin
You previously expressed interest in the addition of this feature. Please reach out if you are interested in collaborating or contributing in any way. Thank you!

**Comment by neilrobert** [comment]
> Hi all! @likeshumidity and I are planning to begin work on this issue. Specifically, we are planning to start with feature D (Save chart or table as SVG) as outlined in the prior [comment](https://github.com/metabase/metabase/issues/4701#issuecomment-675496040).
> 
> This will be our first contribution to Metabase - as such, we are open to and are actively looking for support in whatever manner it is willing to be provided i.e. strategy, development, etc.
> 
> @mijamo @dncpax @alexfrench @blongden @aaugustin You previously expressed interest in the addition of this feature. Please reach out if you are interested in collaborating or contributing in any way. Thank you!

Thats great news - happy to do some UAT if required thx Neil

**Comment by lutvdh** [comment]
Hi @petervuyk456 & @likeshumidity , please advise on your progress.  This is also becoming a demanding requirement from my business.

**Comment by tiguere** [comment]
hello -- is there any progress on this feature?

**Comment by MchlUh** [comment]
Hi, I just want to add something here : 
Until now, I was able to use the "file" --> "print" --> "save to pdf" function of google chrome to export a whole dashboard, and it was very satisfactory !
Since I upgraded from version 42.4 to 43.2 (big jump), It's not possible anymore, because only the first part of the dashboard will be saved to PDF. I think it has something to do with the upgrade and the new layout.
I hope it's appropriate to put that here, otherwise please tell me and I will open a new issue.

**Comment by flamber** [comment]
@MchlUh Unrelated to this issue. It's already fixed in the upcoming 43.3 #23098

**Comment by aaugustin** [comment]
Thank you so much!!!
